File: be/src/main/java/com/q2k/meditech/service/PaymentService.java
package com.q2k.meditech.service;

import com.q2k.meditech.dto.*;
import com.q2k.meditech.dto.receptionist.EndOfDayReportDTO;
import com.q2k.meditech.dto.receptionist.HourlyRevenueDTO;
import com.q2k.meditech.dto.receptionist.PendingPaymentDTO;
import com.q2k.meditech.dto.receptionist.SendPaymentLinkDTO;

/**
 * Payment Service Interface
 */
public interface PaymentService {

    /**
     * Create payment request for an appointment
     * @param dto Payment creation data
     * @param currentUserId User creating the payment
     * @return Created payment
     */
    PaymentDTO createPayment(PaymentCreateDTO dto, Long currentUserId);

    /**
     * Create a PENDING payment for an appointment (e.g. when patient books).
     * If a payment already exists for the appointment, returns it. No payment method required.
     * @param appointmentId Appointment ID
     * @param processedByUserId User who triggered (e.g. patient who booked)
     * @return Created or existing payment
     */
    PaymentDTO createPaymentForAppointment(Long appointmentId, Long processedByUserId);

    /**
     * Create payment request for a prescription
     * @param dto Prescription payment creation data
     * @param currentUserId User creating the payment
     * @return Created payment
     */
    PaymentDTO createPrescriptionPayment(PrescriptionPaymentCreateDTO dto, Long currentUserId);

    /**
     * Initialize MoMo payment (create order and get pay URL)
     * @param paymentId Payment ID
     * @param dto MoMo init parameters
     * @param currentUserId User initializing
     * @return Payment init result with pay URL
     */
    PaymentInitDTO initMomoPayment(Long paymentId, MomoInitDTO dto, Long currentUserId);

    /**
     * Get QR code for payment (MoMo only)
     * @param paymentId Payment ID
     * @param currentUserId User requesting
     * @return QR code data
     */
    PaymentQrDTO getPaymentQr(Long paymentId, Long currentUserId);

    /**
     * Refresh QR code (re-init if expired)
     * @param paymentId Payment ID
     * @param currentUserId User refreshing
     * @return New payment init result
     */
    PaymentInitDTO refreshQr(Long paymentId, Long currentUserId);

    /**
     * Mark payment as paid with cash
     * @param paymentId Payment ID
     * @param dto Mark cash data
     * @param currentUserId User processing
     * @return Updated payment
     */
    PaymentDTO markPaidCash(Long paymentId, MarkCashDTO dto, Long currentUserId);

    /**
     * Cancel payment request
     * @param paymentId Payment ID
     * @param dto Cancel data
     * @param currentUserId User cancelling
     * @return Updated payment
     */
    PaymentDTO cancelPayment(Long paymentId, CancelPaymentDTO dto, Long currentUserId);

    /**
     * Create a PENDING payment for service orders and initialize MoMo QR payment
     * @param appointmentId Appointment ID
     * @param serviceOrderIds List of service order IDs to pay
     * @param currentUserId User initializing the payment
     * @return Payment init result with QR code URL
     */
    PaymentInitDTO createAndInitMomoForServiceOrders(Long appointmentId, java.util.List<Long> serviceOrderIds, Long currentUserId);

    /**
     * Handle MoMo webhook callback
     * @param webhookDto MoMo webhook data
     * @return Updated payment
     */
    PaymentDTO handleMomoWebhook(MomoWebhookDTO webhookDto);

    /**
     * Handle mock webhook (for testing)
     * @param mockDto Mock webhook data
     * @return Updated payment
     */
    PaymentDTO handleMockWebhook(WebhookMockDTO mockDto);

    /**
     * Get payment by ID
     * @param paymentId Payment ID
     * @return Payment data
     */
    PaymentDTO getPaymentById(Long paymentId);

    /**
     * Get patient's payments with pagination and filters
     * @param patientId Patient ID
     * @param status Filter by status (optional)
     * @param method Filter by payment method (optional)
     * @param from From date (optional, yyyy-MM-dd)
     * @param to To date (optional, yyyy-MM-dd)
     * @param pageNumber Page number (0-based)
     * @param pageSize Page size
     * @return Page of payments
     */
    org.springframework.data.domain.Page<PaymentDTO> getMyPayments(
            Long patientId,
            String status,
            String method,
            String from,
            String to,
            int pageNumber,
            int pageSize
    );

    /**
     * Get payment by ID with ownership check
     * @param paymentId Payment ID
     * @param patientId Patient ID (for ownership check)
     * @return Payment data
     */
    PaymentDTO getPaymentByIdForPatient(Long paymentId, Long patientId);

    /**
     * Get payment QR with ownership check
     * @param paymentId Payment ID
     * @param patientId Patient ID (for ownership check)
     * @return QR code data
     */
    PaymentQrDTO getPaymentQrForPatient(Long paymentId, Long patientId);

    /**
     * Patient initiates MoMo payment (with ownership check)
     */
    PaymentInitDTO initMomoPaymentForPatient(Long paymentId, Long patientId);

    /**
     * Patient cancels own PENDING payment (with ownership check)
     */
    PaymentDTO cancelPaymentForPatient(Long paymentId, Long patientId, String reason);

    // ========== ADMIN METHODS ==========

    /**
     * List all payments with filters (admin)
     * @param status Filter by status (optional)
     * @param method Filter by payment method (optional)
     * @param from From date (optional, yyyy-MM-dd)
     * @param to To date (optional, yyyy-MM-dd)
     * @param patientId Filter by patient ID (optional)
     * @param appointmentId Filter by appointment ID (optional)
     * @param pageNumber Page number (0-based)
     * @param pageSize Page size
     * @return Page of payments
     */
    org.springframework.data.domain.Page<PaymentDTO> getAllPayments(
            String status,
            String method,
            String from,
            String to,
            Long patientId,
            Long appointmentId,
            int pageNumber,
            int pageSize
    );

    /**
     * Refund payment (cash or MoMo)
     * @param paymentId Payment ID
     * @param dto Refund parameters
     * @param currentUserId User processing refund
     * @return Updated payment
     */
    PaymentDTO refundPayment(Long paymentId, RefundDTO dto, Long currentUserId);

    /**
     * Cancel payment (admin can cancel any payment)
     * @param paymentId Payment ID
     * @param dto Cancel data
     * @param currentUserId User cancelling
     * @return Updated payment
     */
    PaymentDTO adminCancelPayment(Long paymentId, CancelPaymentDTO dto, Long currentUserId);

    /**
     * Expire pending/initiated payments (scheduled job)
     * @param dto Expiry parameters
     * @return Expiry result
     */
    ExpirePaymentsResultDTO expirePayments(ExpirePaymentsDTO dto);

    /**
     * Reconcile MoMo payment status
     * @param paymentId Payment ID
     * @param currentUserId User initiating reconciliation
     * @return Updated payment
     */
    PaymentDTO reconcileMomoStatus(Long paymentId, Long currentUserId);

    // ========== ADVANCED ADMIN METHODS ==========

    /**
     * List all payments with advanced filters (admin)
     * @param search Search by transaction code, patient name, appointment code
     * @param status Filter by status (optional)
     * @param method Filter by payment method (optional)
     * @param doctorId Filter by doctor ID (optional)
     * @param patientId Filter by patient ID (optional)
     * @param minAmount Minimum amount filter (optional)
     * @param maxAmount Maximum amount filter (optional)
     * @param from From date (optional, yyyy-MM-dd)
     * @param to To date (optional, yyyy-MM-dd)
     * @param pageNumber Page number (0-based)
     * @param pageSize Page size
     * @param sortBy Sort field (default: paymentDate)
     * @param sortDir Sort direction (default: DESC)
     * @return Page of payments
     */
    org.springframework.data.domain.Page<PaymentDTO> getAllPaymentsAdvanced(
            String search,
            String status,
            String method,
            Long doctorId,
            Long patientId,
            java.math.BigDecimal minAmount,
            java.math.BigDecimal maxAmount,
            String from,
            String to,
            int pageNumber,
            int pageSize,
            String sortBy,
            String sortDir
    );

    /**
     * Get payment statistics for dashboard
     * @param from From date (optional, yyyy-MM-dd)
     * @param to To date (optional, yyyy-MM-dd)
     * @return Payment statistics
     */
    PaymentStatsDTO getPaymentStatistics(String from, String to);

    /**
     * Bulk mark payments as paid
     * @param dto Bulk mark paid data
     * @param currentUserId User processing
     * @return Bulk action result
     */
    PaymentBulkResultDTO bulkMarkAsPaid(BulkMarkPaidDTO dto, Long currentUserId);

    /**
     * Export payments to file
     * @param search Search filter
     * @param status Status filter
     * @param method Payment method filter
     * @param doctorId Doctor ID filter
     * @param patientId Patient ID filter
     * @param minAmount Minimum amount filter
     * @param maxAmount Maximum amount filter
     * @param from From date filter
     * @param to To date filter
     * @param format Export format (EXCEL, CSV, PDF)
     * @return File bytes
     */
    byte[] exportPayments(
            String search,
            String status,
            String method,
            Long doctorId,
            Long patientId,
            java.math.BigDecimal minAmount,
            java.math.BigDecimal maxAmount,
            String from,
            String to,
            String format
    );

    // ========== PAYMENT DETAIL & ACTIONS APIs ==========

    /**
     * Get comprehensive payment details (admin)
     * @param paymentId Payment ID
     * @return Payment detail with all related info
     */
    PaymentDetailDTO getPaymentDetail(Long paymentId);

    /**
     * Mark payment as paid manually (for cash/offline payments)
     * @param paymentId Payment ID
     * @param dto Mark paid data
     * @param currentUserId User processing
     * @return Updated payment
     */
    PaymentDTO markAsPaid(Long paymentId, MarkPaidDTO dto, Long currentUserId);

    /**
     * Retry payment - resend payment link to patient
     * @param paymentId Payment ID
     * @param dto Retry data containing send method
     * @param currentUserId User initiating retry
     * @return Message result
     */
    MessageDTO retryPayment(Long paymentId, RetryPaymentDTO dto, Long currentUserId);

    /**
     * Send payment receipt to email
     * @param paymentId Payment ID
     * @param dto Send receipt data
     * @param currentUserId User sending receipt
     * @return Message result
     */
    MessageDTO sendReceipt(Long paymentId, SendReceiptDTO dto, Long currentUserId);

    /**
     * Generate and get receipt PDF as bytes
     * @param paymentId Payment ID
     * @return PDF bytes
     */
    byte[] downloadReceipt(Long paymentId);

    /**
     * Get payment history/timeline
     * @param paymentId Payment ID
     * @return Payment history with events
     */
    PaymentHistoryDTO getPaymentHistory(Long paymentId);

    // ========== RECEPTIONIST TAB 5 — NEW METHODS ==========

    /**
     * Get pending payments list (appointment COMPLETED + payment PENDING)
     * Tab 5.3 — Pending debts
     */
    org.springframework.data.domain.Page<PendingPaymentDTO> getPendingPayments(
            String search, int pageNumber, int pageSize, String sortBy, String sortDir);

    /**
     * Send MoMo payment link to patient via SMS/email
     * Tab 5.3 — Pending actions
     */
    MessageDTO sendPaymentLink(Long paymentId, SendPaymentLinkDTO dto, Long currentUserId);

    /**
     * Get hourly revenue data for today (for chart)
     * Tab 5.4 — Today's Revenue
     */
    HourlyRevenueDTO getHourlyRevenue(String date);

    /**
     * Generate end-of-day report
     * Tab 5.4 — Today's Revenue
     */
    EndOfDayReportDTO generateEndOfDayReport(String date, Long currentUserId);

    /**
     * Export end-of-day report as PDF or Excel bytes
     */
    byte[] exportEndOfDayReport(String date, String format, Long currentUserId);
}

File: fe/src/services/paymentService.ts
import api from './api';

// ==================== TYPES ====================

export interface PaymentDTO {
  id: number;
  transactionCode: string;
  paymentDate: string;
  patientName: string;
  doctorName: string;
  appointmentCode: string;
  amount: number;
  totalAmount: number;
  paymentMethod: string;
  status: 'PENDING' | 'PAID' | 'INITIATED' | 'PROCESSING' | 'FAILED' | 'REFUNDED' | 'CANCELLED' | 'EXPIRED';
  paymentStatus: string;
  currency: string;
  prescriptionId?: number;
  prescriptionCode?: string;
  referenceType?: 'APPOINTMENT' | 'PRESCRIPTION';
}

export interface PaymentDetailDTO {
  id: number;
  transactionCode: string;
  paymentCode: string;
  status: 'PENDING' | 'PAID' | 'INITIATED' | 'PROCESSING' | 'FAILED' | 'REFUNDED' | 'CANCELLED' | 'EXPIRED';
  paymentStatus: string;
  paymentDate: string;
  createdDate: string;
  createdAt: string;
  lastUpdated: string;
  updatedAt: string;
  amount: number;
  discountAmount: number;
  taxAmount: number;
  totalAmount: number;
  currency: string;
  paymentMethod: string;
  transactionFee: number;
  netAmount: number;
  tax: number;
  patientName: string;
  patientEmail: string;
  patientPhone: string;
  patientId: number;
  doctorName: string;
  doctorSpecialization: string;
  doctorId: number;
  appointmentCode: string;
  appointmentDate: string;
  appointmentType: string;
  appointmentStatus: string;
  appointmentId: number;
  gatewayName?: string;
  gatewayTransactionId?: string;
  gatewayStatus?: string;
  gatewayResponse?: string;
  isRefunded?: boolean;
  refundAmount?: number;
  refundDate?: string;
  refundReason?: string;
  refundedBy?: string;
  refundedByName?: string;
  refundStatus?: string;
  processedBy?: number;
  processedByName?: string;
  notes?: string;
  adminNotes?: string;
  patientNotes?: string;
  invoiceId?: number;
  invoiceNumber?: string;
  invoiceStatus?: string;
}

export interface PaymentStatsDTO {
  // Today's revenue
  todayRevenue: number;
  todayTransactionCount: number;
  todayAverageTransaction: number;
  // This month's revenue
  monthRevenue: number;
  lastMonthRevenue: number;
  monthGrowthPercentage: number;
  monthTransactionCount: number;
  // Payment methods distribution
  paymentMethodsDistribution: {
    method: string;
    count: number;
    amount: number;
    percentage: number;
  }[];
  // Pending payments
  pendingCount: number;
  pendingAmount: number;
  // Refund statistics
  totalRefundedThisMonth: number;
  refundCountThisMonth: number;
  refundRatePercentage: number;
  // Status breakdown
  statusCounts: Record<string, number>;
  // Date range
  fromDate: string;
  toDate: string;
}

export interface PaymentHistoryDTO {
  paymentId?: number;
  paymentCode?: string;
  events: {
    id?: number;
    eventType: string;
    event: string;
    description: string;
    previousValue?: string;
    newValue?: string;
    performedByName?: string;
    performedById?: number;
    eventTime: string;
    timestamp: string;
    user?: string;
    metadata?: any;
  }[];
}

export interface PaymentBulkResultDTO {
  success: number;
  failed: number;
  errors?: string[];
}

export interface BulkMarkPaidDTO {
  paymentIds: number[];
  paymentMethod: string;
  notes?: string;
}

export interface MarkPaidDTO {
  paymentMethod?: string;
  notes?: string;
}

export interface RefundDTO {
  refundAmount: number;
  refundReason: string;
  notes?: string;
  transactionId?: string;
  description?: string;
}

export interface CancelPaymentDTO {
  reason: string;
  notes?: string;
}

export interface RetryPaymentDTO {
  sendVia: 'EMAIL' | 'SMS';
}

export interface SendReceiptDTO {
  email: string;
}

export interface MessageDTO {
  message: string;
  success: boolean;
}

export interface PaymentFilterParams {
  search?: string;
  status?: string;
  paymentMethod?: string;
  doctorId?: number;
  patientId?: number;
  minAmount?: number;
  maxAmount?: number;
  from?: string;
  to?: string;
  pageNumber?: number;
  pageSize?: number;
  sortBy?: string;
  sortDir?: string;
}

/** Extension for downloaded file (backend: EXCEL → XLSX, CSV, PDF only). */
export function paymentExportFileExtension(format: string): string {
  switch (format.toUpperCase()) {
    case 'CSV':
      return 'csv';
    case 'PDF':
      return 'pdf';
    case 'EXCEL':
    default:
      return 'xlsx';
  }
}

async function readBlobErrorMessage(blob: Blob): Promise<string> {
  try {
    const text = await blob.text();
    if (!text?.trim()) return 'Export failed';
    try {
      const j = JSON.parse(text) as { message?: string; error?: string; detail?: string };
      return j.message || j.error || (typeof j.detail === 'string' ? j.detail : '') || text.slice(0, 300);
    } catch {
      return text.slice(0, 300);
    }
  } catch {
    return 'Export failed';
  }
}

export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

// ==================== API FUNCTIONS ====================

/**
 * Get all payments with advanced filters
 * GET /api/admin/payments
 */
export const getAllPayments = async (
  params: PaymentFilterParams
): Promise<PageResponse<PaymentDTO>> => {
  const response = await api.get('/admin/payments', { params });
  return response.data;
};

/**
 * Get payment statistics for dashboard
 * GET /api/admin/payments/statistics
 */
export const getPaymentStatistics = async (
  from?: string,
  to?: string
): Promise<PaymentStatsDTO> => {
  const response = await api.get('/admin/payments/statistics', {
    params: { from, to },
  });
  return response.data;
};

/**
 * Bulk mark multiple pending payments as paid
 * POST /api/admin/payments/bulk/mark-paid
 */
export const bulkMarkAsPaid = async (
  dto: BulkMarkPaidDTO
): Promise<PaymentBulkResultDTO> => {
  const response = await api.post('/admin/payments/bulk/mark-paid', dto);
  return response.data;
};

/**
 * Export filtered transactions to file
 * GET /api/admin/payments/export
 */
export const exportPayments = async (
  params: PaymentFilterParams & { format: string }
): Promise<Blob> => {
  try {
    const response = await api.get('/admin/payments/export', {
      params,
      responseType: 'blob',
    });
    const blob = response.data as Blob;
    // Some error responses are JSON with responseType blob
    if (blob.type?.includes('application/json')) {
      throw new Error(await readBlobErrorMessage(blob));
    }
    return blob;
  } catch (err: unknown) {
    const ax = err as { response?: { data?: unknown } };
    const data = ax?.response?.data;
    if (data instanceof Blob) {
      throw new Error(await readBlobErrorMessage(data));
    }
    throw err;
  }
};

/**
 * Get comprehensive payment detail
 * GET /api/admin/payments/{id}
 */
export const getPaymentDetail = async (id: number): Promise<PaymentDetailDTO> => {
  const response = await api.get(`/admin/payments/${id}`);
  return response.data;
};

/**
 * Manually mark payment as paid
 * POST /api/admin/payments/{id}/mark-paid
 */
export const markPaymentAsPaid = async (
  id: number,
  dto: MarkPaidDTO
): Promise<PaymentDTO> => {
  const response = await api.post(`/admin/payments/${id}/mark-paid`, dto);
  return response.data;
};

/**
 * Retry payment - resend payment link
 * POST /api/admin/payments/{id}/retry
 */
export const retryPayment = async (
  id: number,
  dto: RetryPaymentDTO
): Promise<MessageDTO> => {
  const response = await api.post(`/admin/payments/${id}/retry`, dto);
  return response.data;
};

/**
 * Send payment receipt to patient
 * POST /api/admin/payments/{id}/send-receipt
 */
export const sendReceipt = async (
  id: number,
  dto: SendReceiptDTO
): Promise<MessageDTO> => {
  const response = await api.post(`/admin/payments/${id}/send-receipt`, dto);
  return response.data;
};

/**
 * Download payment receipt as PDF
 * GET /api/admin/payments/{id}/receipt
 */
export const downloadReceipt = async (id: number): Promise<Blob> => {
  const response = await api.get(`/admin/payments/${id}/receipt`, {
    responseType: 'blob',
  });
  return response.data;
};

/**
 * Get payment history/timeline
 * GET /api/admin/payments/{id}/history
 */
export const getPaymentHistory = async (id: number): Promise<PaymentHistoryDTO> => {
  const response = await api.get(`/admin/payments/${id}/history`);
  return response.data;
};

/**
 * Refund payment (cash or MoMo)
 * POST /api/admin/payments/{id}/refund
 */
export const refundPayment = async (
  id: number,
  dto: RefundDTO
): Promise<PaymentDTO> => {
  const response = await api.post(`/admin/payments/${id}/refund`, dto);
  return response.data;
};

/**
 * Cancel payment (admin override)
 * PATCH /api/admin/payments/{id}/cancel
 */
export const cancelPayment = async (
  id: number,
  dto: CancelPaymentDTO
): Promise<PaymentDTO> => {
  const response = await api.patch(`/admin/payments/${id}/cancel`, dto);
  return response.data;
};

/**
 * Reconcile MoMo payment status
 * POST /api/admin/payments/{id}/reconcile/momo
 */
export const reconcileMomoStatus = async (id: number): Promise<PaymentDTO> => {
  const response = await api.post(`/admin/payments/${id}/reconcile/momo`);
  return response.data;
};

// ==================== HELPER FUNCTIONS ====================

/**
 * Download a file from blob data
 */
export const downloadFile = (blob: Blob, filename: string) => {
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.setAttribute('download', filename);
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(url);
};

/**
 * Format currency (VND)
 */
export const formatCurrency = (amount: number, currency: string = 'VND') => {
  return new Intl.NumberFormat('vi-VN', {
    style: 'currency',
    currency: currency,
  }).format(amount);
};

/**
 * Badge color type (matching Badge component)
 */
export type BadgeColor =
  | "primary"
  | "success"
  | "error"
  | "warning"
  | "info"
  | "light"
  | "dark";

/**
 * Get status badge color
 */
export const getStatusColor = (status: string): BadgeColor => {
  switch (status) {
    case 'PENDING':
      return 'warning';
    case 'PAID':
      return 'success';
    case 'INITIATED':
    case 'PROCESSING':
      return 'info';
    case 'FAILED':
      return 'error';
    case 'REFUNDED':
      return 'info';
    case 'CANCELLED':
    case 'EXPIRED':
      return 'light';
    default:
      return 'light';
  }
};

/**
 * Get payment method icon
 */
export const getPaymentMethodIcon = (method: string): string => {
  switch (method) {
    case 'CASH':
      return '💵';
    case 'MOMO':
      return '📱';
    default:
      return '💳';
  }
};

/**
 * Get payment method label
 */
export const getPaymentMethodLabel = (method: string): string => {
  const labels: Record<string, string> = {
    CASH: 'Cash',
    CARD: 'Card',
    BANK_TRANSFER: 'Bank Transfer',
    E_WALLET: 'E-Wallet',
    INSURANCE: 'Insurance',
    MOMO: 'MoMo',
    ZALOPAY: 'ZaloPay',
    VNPAY: 'VNPay',
  };
  return labels[method] || method;
};

// ==================== RECEPTIONIST PAYMENT API ====================

export interface ReceptionistPaymentCreateDTO {
  appointmentId: number;
  paymentMethod: 'CASH' | 'MOMO';
  discountAmount?: number;
  taxAmount?: number;
  notes?: string;
}

export interface PrescriptionPaymentCreateDTO {
  prescriptionId: number;
  paymentMethod: 'CASH' | 'MOMO';
  discountAmount?: number;
  taxAmount?: number;
  notes?: string;
}

export interface MarkCashDTO {
  amountReceived?: number;
  changeGiven?: number;
  transactionId?: string;
  notes?: string;
  printReceipt?: boolean;
  emailReceipt?: boolean;
  smsReceipt?: boolean;
}

export interface PaymentInitResult {
  paymentId: number;
  paymentCode: string;
  payUrl: string;
  qrCodeUrl: string;
  orderId: string;
  message: string;
  success: boolean;
}

export interface PaymentQrResult {
  id: number;
  paymentId: number;
  provider: string;
  qrPayload: string;
  qrDataUrl?: string;
  payUrl: string;
  expiresAt: string;
  status: string;
  createdAt: string;
}

/**
 * Create payment for an appointment (receptionist)
 * POST /api/receptionist/payments
 */
export const createReceptionistPayment = async (dto: ReceptionistPaymentCreateDTO): Promise<PaymentDTO> => {
  const response = await api.post('/receptionist/payments', dto);
  return response.data;
};

/**
 * Create prescription payment (receptionist)
 * POST /api/receptionist/payments/prescription
 */
export const createPrescriptionPayment = async (dto: PrescriptionPaymentCreateDTO): Promise<PaymentDTO> => {
  const response = await api.post('/receptionist/payments/prescription', dto);
  return response.data;
};

/**
 * Initialize MoMo payment (receptionist)
 * POST /api/receptionist/payments/{id}/momo/init
 */
export const initMomoPayment = async (paymentId: number): Promise<PaymentInitResult> => {
  const response = await api.post(`/receptionist/payments/${paymentId}/momo/init`, {});
  return response.data;
};

/**
 * Get QR code for payment (receptionist)
 * GET /api/receptionist/payments/{id}/qr
 */
export const getPaymentQr = async (paymentId: number): Promise<PaymentQrResult> => {
  const response = await api.get(`/receptionist/payments/${paymentId}/qr`);
  return response.data;
};

/**
 * Refresh QR code (receptionist)
 * POST /api/receptionist/payments/{id}/qr/refresh
 */
export const refreshPaymentQr = async (paymentId: number): Promise<PaymentInitResult> => {
  const response = await api.post(`/receptionist/payments/${paymentId}/qr/refresh`);
  return response.data;
};

/**
 * Mark payment as paid with cash (receptionist)
 * PATCH /api/receptionist/payments/{id}/mark-cash
 */
export const markPaymentCash = async (paymentId: number, dto?: MarkCashDTO): Promise<PaymentDTO> => {
  const response = await api.patch(`/receptionist/payments/${paymentId}/mark-cash`, dto || {});
  return response.data;
};

/**
 * Cancel payment (receptionist)
 * PATCH /api/receptionist/payments/{id}/cancel
 */
export const cancelReceptionistPayment = async (paymentId: number, reason?: string): Promise<PaymentDTO> => {
  const response = await api.patch(`/receptionist/payments/${paymentId}/cancel`, { reason: reason || 'Cancelled by receptionist' });
  return response.data;
};

/**
 * Get payment by ID (receptionist)
 * GET /api/receptionist/payments/{id}
 */
export const getReceptionistPayment = async (paymentId: number): Promise<PaymentDTO> => {
  const response = await api.get(`/receptionist/payments/${paymentId}`);
  return response.data;
};

export default {
  getAllPayments,
  getPaymentStatistics,
  bulkMarkAsPaid,
  paymentExportFileExtension,
  exportPayments,
  getPaymentDetail,
  markPaymentAsPaid,
  retryPayment,
  sendReceipt,
  downloadReceipt,
  getPaymentHistory,
  refundPayment,
  cancelPayment,
  reconcileMomoStatus,
  createReceptionistPayment,
  createPrescriptionPayment,
  initMomoPayment,
  getPaymentQr,
  refreshPaymentQr,
  markPaymentCash,
  cancelReceptionistPayment,
  getReceptionistPayment,
  downloadFile,
  formatCurrency,
  getStatusColor,
  getPaymentMethodIcon,
  getPaymentMethodLabel,
};
